### PR TITLE
Add possibility of modifying dynamic objects behavior 

### PR DESCRIPTION
**Pull request by forkerer**
This pull request aims to add possibility of modifying behavior of dynamic objects. This includes properties like their mass, particle effects played by them, if their explode or not, etc. By now mta has setObjectProperty() function implemented, but it's limited to custom created objects, and doesn't allow to change behavior of original world objects. All default properties, along with simple documentation, can be found in object.dat in GTA SA data folder. All changed properties are restored to default once user leaves server.

# 1. Property groups used by certain models IDs
SA by default contains 160 property groups for dynamic objects, multiple models can share single properties group, i've added possibility to change which group is used by which model, which required addition of those 3 functions:
```lua
-- Returns properties group used by given model ID, will return 0 if model isn't dynamic, as it's defaulted to that in engine
engineGetModelPhysicalPropertiesGroup(modelID) 

-- Changes which property group given model will use
engineSetModelPhysicalPropertiesGroup(modelID, groupID)

-- Restores model to default state
engineRestoreModelPhysicalPropertiesGroup(modelID)
```
I didn't notice any change when i've changed group id's of static objects (or any non dynamic stuff in CModelInfo table), but game is stable when those are changed for them, so i didn't disable ability to do so.

# 2. Properties themselves
There are multiple adjustable properties for each group, most of the changes made in them are visible after object is created, so most changes require you to get far away from object, and back to become visible.
Those 3 functions were created to give ability to modify those properties:
```lua
-- Property setters
engineSetObjectGroupPhysicalProperty(groupID, objectgroup-modifiable property, [...args] )

-- Property getters
engineGetObjectGroupPhysicalProperty(groupID, objectgroup-modifiable property)

-- Used to restore default property values
engineRestoreObjectGroupPhysicalProperties(groupID)
```

Those are the modifiable properties, along with their lua names and types:
```cpp
enum eObjectGroup::Modifiable
{
    MASS,                //mass                         float
    TURNMASS,            //turn_mass                    float
    AIRRESISTANCE,       //air_resistance               float
    ELASTICITY,          //elasticity                   float
    BUOYANCY,            //buoyancy                     float
    UPROOTLIMIT,         //uproot_limit                 float
    COLDAMAGEMULTIPLIER, //col_damage_multiplier        float
    COLDAMAGEEFFECT,     //col_damage_effect            eObjectGroup::DamageEffect
    SPECIALCOLRESPONSE,  //special_col_response         eObjectGroup::CollisionResponse
    CAMERAAVOID,         //avoid_camera                 bool
    EXPLOSION,           //cause_explosion              bool
    FXTYPE,              //fx_type                      eObjectGroup::FxType
    FXOFFSET,            //fx_offset                    CVector
    FXSYSTEM,            //fx_system                    takes string, uses FxSystemBlueprint
    SMASHMULTIPLIER,     //smash_multiplier             float
    BREAKVELOCITY,       //break_velocity               CVector
    BREAKVELOCITYRAND,   //break_velocity_randomness    float
    BREAKMODE,           //break_mode                   eObjectGroup::BreakMode
    SPARKSONIMPACT       //sparks_on_impact             bool
};
```
The specialized properties enums looks like so:
```c
enum DamageEffect // Dictates which damage effect is applied to object on collision
{
    NO_EFFECT = 0,          //none              Object doesn't change at all once it's damaged
    CHANGE_MODEL = 1,       //change_model      Some of the objects change model on collision, those use this
    SMASH_COMPLETELY = 20,  //smash             Object is smashed
    CHANGE_THEN_SMASH = 21, //change_smash      First CHANGE_MODEL, afterwards smash on collision
    BREAKABLE = 200,        //breakable         Object is breakable normally
    BREAKABLE_REMOVED = 202 //breakable_remove  object.dat says: '(ie. never regenerated after destroyed)'
};

enum CollisionResponse // How object responds to being collided with
{
    NO_RESPONSE, //none
    LAMPPOST,    //lamppost
    SMALLBOX,    //small_box
    BIGBOX,      //big_box
    FENCEPART,   //fence_part
    GRENADE,     //grenade
    SWINGDOOR,   //swingdoor
    LOCKDOOR,    //lockdoor
    HANGING,     //hanging
    POOLBALL     //poolball
};

enum FxType // Dictates when particles will be created when colliding with object
{
    NO_FX,                //none
    PLAY_ON_HIT,          //play_on_hit
    PLAY_ON_DESTROYED,    //play_on_destroyed
    PLAY_ON_HIT_DESTROYED //play_on_hitdestroyed
};

enum BreakMode // Dictates how object can be damaged, info from object.dat:
{
    NOT_BY_GUN, //not_by_gun    'not breakable by gun'
    BY_GUN,     //by_gun        'breakable'
    SMASHABLE,  //smashable     'smashable'
};
```
### Fx system:
Any particle system included in SA could be used by dynamic objects, but after testing i noticed that looping particle systems stay on map forever (even after leaving server), so i had to disable ability to use those, only non looping system are acceptable by fx_system property, those are: (from effects.fxp file)
```
wallbust,
shootlight,
puke,
explosion_door,
explosion_crate,
explosion_barrel,
blood_heli,
tree_hit_palm,
tree_hit_fir,
water_swim,
water_splsh_sml,
water_splash_big,
water_splash,
water_hydrant,
tank_fire,
riot_smoke,
gunsmoke,
gunflash,
explosion_tiny,
explosion_small,
explosion_molotov,
explosion_medium,
explosion_large,
explosion_fuel_car,
exhale,
camflash,
prt_wake
```

# 3. Test resource
DOWNLOAD: [props.zip](https://github.com/multitheftauto/mtasa-blue/files/2776563/props.zip)
I'm including simple resource containing editor for those properties, that looks like that:

image

Simply start resource, and write /showgui to make it visible, same command to hide it back again.

Properties editor:
-- Changes properties of given group
Model editor:
-- Gives ability to change properties group used by given model
Presets:
-- Some simple presets, and ability to restore properties of all groups.
-- Given presets are: make everything explode, make everything easy to destroy, make everything undestroyable

### CODE REVIEW

**Comment by Einheit-101** [comment]
avoid_camera is interesting, is it possible to set this property to projectile objects? Because its highly annoying that projectiles interact with the camera.

**Comment by qaisjp** [comment]
Lua API seems reasonable. Is it possible to create additional object groups or is it limited to the default set of groups?

**Comment by forkerer** [comment]
> Lua API seems reasonable. Is it possible to create additional object groups or is it limited to the default set of groups?

Limited to default 160 groups, dunno how many references to current table location are in the gta sa, and current amount is plenty for what's needed i think.

File: Client/game_sa/CObjectGroupPhysicalPropertiesSA.cpp [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+#include "CObjectGroupPhysicalPropertiesSA.h"
+extern CGameSA* pGame;
+
+CObjectGroupPhysicalPropertiesSAInterface* pObjectInfo = (CObjectGroupPhysicalPropertiesSAInterface*)ARRAY_ObjectGroupsDynamicInfo;
+std::unordered_map<unsigned char, CObjectGroupPhysicalPropertiesSAInterface*> CObjectGroupPhysicalPropertiesSA::ms_OriginalGroupProperties;
+
+CObjectGroupPhysicalPropertiesSA::CObjectGroupPhysicalPropertiesSA()
+{
+    m_pInterface = nullptr;
+}
```
Comment (patrikjuvonen) [inline Client/game_sa/CObjectGroupPhysicalPropertiesSA.cpp]:
```cpp
CObjectGroupPhysicalPropertiesSA::CObjectGroupPhysicalPropertiesSA() : m_pInterface(nullptr) {}
```
Prefer initializer list here instead [[1]](https://www.geeksforgeeks.org/when-do-we-use-initializer-list-in-c/)

File: Client/game_sa/CObjectGroupPhysicalPropertiesSA.cpp [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+    if (fabs(m_pInterface->fAirResistance - fAirResistance) < FLOAT_EPSILON)
+        return;
+
+    ChangeSafeguard();
+    m_pInterface->fAirResistance = fAirResistance;
+}
+
+float CObjectGroupPhysicalPropertiesSA::GetAirResistance()
+{
+    ChangeSafeguard();
```
Comment (patrikjuvonen) [inline Client/game_sa/CObjectGroupPhysicalPropertiesSA.cpp]: 1. Why is `ChangeSafeguard();` called here, but not consistently across all getters?
2. Maybe also change all applicable member functions to `const`? [[1]](https://www.geeksforgeeks.org/const-member-functions-c/)

**Review by patrikjuvonen — CHANGES_REQUESTED** [review]

**Review by patrikjuvonen — APPROVED** [review]
Thanks for the changes. Looks good to me!

I'm looking at this line: https://github.com/multitheftauto/mtasa-blue/pull/784/files?w=1#diff-72a39fa83b01cb9bbbfc41c8928ac793R1343 and right now I have PR #299 open for current LOD functions that crash the client if passed an ID of 20000+. You seem to error out if the ID is not between 0 and 20000. Can you give a little more context on why this is? I'm trying to understand what's beyond 19999 that makes it crash the client so I can figure out a permanent solution with good context for the other issue.

@qaisjp commented:
> If there is an underlying bug with `CGameSA::GetModelInfo` then that should be fixed instead.

I'm trying to figure out how to give closure to this 20000 issue.

**Comment by forkerer** [comment]
As far as I'm aware, (and in the IDB i'm using), 20000 is original size of that array in gta sa, trying to change physical properties of anything above that crashes the client. I think the additional space reserved in CGameSA (up to 26000) was made either by mistake, or in case of mta team wanting to expand some limits.

**Comment by patrikjuvonen** [comment]
Found something old but interesting regarding dynamic objects on the Wiki: https://wiki.multitheftauto.com/wiki/Dynamic_object

File: Client/game_sa/CFxManagerSA.cpp [hunk 3]
```diff
@@ -67,6 +67,32 @@ void CFxManagerSA::OnFxSystemSAInterfaceDestroyed(CFxSystemSAInterface* pFxSyste
...
 }
 
+CFxSystemBPSAInterface* CFxManagerSA::GetFxSystemBlueprintByName(SString sName)
+{
+    DWORD dwThis = (DWORD)m_pInterface;
+    DWORD dwFunc = FUNC_FxManager_c__GetSystemByName;
+    const char* pChars = sName;
+    CFxSystemBPSAInterface* pFxSystemBlueprint = nullptr;
+
+    _asm
```
Comment (sbx320) [inline Client/game_sa/CFxManagerSA.cpp]: Just call the function directly. _asm is usually not needed, breaks optimizations and just causes a lot of clutter.
```suggestion
    using func_t = (CFxSystemBPSAInterface*)(__thiscall*)(CFxManagerSAInterface* pInterface, const char* szName);
    auto func = reinterpret_cast<func_t>(FUNC_FxManager_c__GetSystemByName);
    return func(m_pInterface, sName);
```

File: Client/game_sa/CFxManagerSA.cpp [hunk 4]
```diff
@@ -67,6 +67,32 @@ void CFxManagerSA::OnFxSystemSAInterfaceDestroyed(CFxSystemSAInterface* pFxSyste
...
+    {
+        mov     ecx, dwThis 
+        push    pChars
+        call    dwFunc
+        mov     pFxSystemBlueprint, eax
+    }
+
+    if (!pFxSystemBlueprint)
```
Comment (sbx320) [inline Client/game_sa/CFxManagerSA.cpp]: Not needed with the other change, but just for the future:
```cpp
if(!pointer) return nullptr;
return pointer;
```
is the same as 
```
return pointer;
```

File: Client/game_sa/CModelInfoSA.cpp [hunk 5]
```diff
@@ -1367,6 +1368,46 @@ void CModelInfoSA::ResetSupportedUpgrades()
...
+    if (!MapFind(ms_OriginalObjectPropertiesGroups, m_dwModelID))
+        MapSet(ms_OriginalObjectPropertiesGroups, m_dwModelID, usOrgGroup);
+
+    GetInterface()->usDynamicIndex = usNewGroup;
+}
+
+unsigned short CModelInfoSA::GetObjectPropertiesGroup()
+{
+    unsigned short usGroup = GetInterface()->usDynamicIndex;
+    if (usGroup == 0xFFFF)
```
Comment (sbx320) [inline Client/game_sa/CModelInfoSA.cpp:1392]: What is the meaning of a `0xFFFF (-1)` property group?

File: Client/game_sa/CModelInfoSA.cpp [hunk 6]
```diff
@@ -1367,6 +1368,46 @@ void CModelInfoSA::ResetSupportedUpgrades()
...
+    if (usGroupInMap)
+    {
+        GetInterface()->usDynamicIndex = *usGroupInMap;
+        MapRemove(ms_OriginalObjectPropertiesGroups, m_dwModelID);
+    }
+}
+
+void CModelInfoSA::RestoreAllObjectsPropertiesGroups()
+{
+    for (auto& pair : ms_OriginalObjectPropertiesGroups)
```
Comment (sbx320) [inline Client/game_sa/CModelInfoSA.cpp]:
```suggestion
    for (const auto& pair : ms_OriginalObjectPropertiesGroups)
```

File: Client/game_sa/CObjectGroupPhysicalPropertiesSA.cpp [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+void CObjectGroupPhysicalPropertiesSA::ChangeSafeguard()
+{
+    if (m_bModified)
+        return;
+
+    m_bModified = true;
+    // Make copy of original
+    if (!MapFind(ms_OriginalGroupProperties, m_ucObjectGroup))
+    {
+        auto pOriginalCopy = new CObjectGroupPhysicalPropertiesSAInterface;
```
Comment (sbx320) [inline Client/game_sa/CObjectGroupPhysicalPropertiesSA.cpp]: this should really be a `std::unique_ptr`, as it's rather hard to follow where the memory is freed.

File: Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp [hunk 8]
```diff
@@ -1317,3 +1329,486 @@ int CLuaEngineDefs::EngineResetSurfaceProperties(lua_State* luaVM)
...
+        {
+            argStream.SetCustomError("Expected model ID in range [0-19999] at argument 1");
+            m_pScriptDebugging->LogCustom(luaVM, argStream.GetFullErrorMessage());
+            lua_pushnil(luaVM);
+            return 1;
+        }
+
+        if (iNewGroup < 0 || iNewGroup > 159)
+        {
+            argStream.SetCustomError("Expected group ID in range [0-159] at argument 1");
```
Comment (sbx320) [inline Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp:1392]: doesn't quite line up with the error above (0 - 19999 includes 19999, 0 - 159 excludes 159?

**Review by sbx320 — CHANGES_REQUESTED** [review]
Looks good in general. I added some comments, but other than that it should be good.

Only issue apart from that would be the discussion on changing result on usage errors in scripting #821

File: Client/game_sa/CModelInfoSA.cpp (hunk 5 above)
Comment (forkerer) [reply Client/game_sa/CModelInfoSA.cpp:1392]: If i remember correctly, it's used by the game to determine that this object id doesn't have any custom physical properties assigned to it, in such cases the default(0) physical properties group is used for most physical calculations on those objects

File: Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp (hunk 8 above)
Comment (forkerer) [reply Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp:1392]: Changed the upper check to match the format, and now it check against 19999

**Comment by ghost** [comment]
Nice work! Does the collision properties work for custom objects? I'm planning to make `CColModel` object based rather than model based.

**Comment by forkerer** [comment]
Since most of the properties are assigned in object constructor, based on model group ID, it should work without problems on custom models that replace dynamic objects. I haven't tested that tho so can't be sure.

**Review by sbx320 — APPROVED** [review]
Looks good. Just needs a decision on the script error thing...

**Comment by ghost** [comment]
Once this PR is merged, modelers can use this to generate breakable models: https://github.com/DK22Pac/v2saconv

File: Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp [hunk 9]
```diff
@@ -1328,3 +1340,486 @@ int CLuaEngineDefs::EngineResetSurfaceProperties(lua_State* luaVM)
...
+
+int CLuaEngineDefs::EngineGetModelPhysicalPropertiesGroup(lua_State* luaVM)
+{
+    //  int engineGetModelPhysicalPropertiesGroup ( int modelID )
+    int iModelID;
+
+    CScriptArgReader argStream(luaVM);
+    argStream.ReadNumber(iModelID);
+
+    if (!argStream.HasErrors())
```
Comment (patrikjuvonen) [inline Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp:1352]: These errors should be changed accordingly (https://github.com/multitheftauto/mtasa-blue/issues/821#issuecomment-522187565).

File: Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp [hunk 10]
```diff
@@ -1328,3 +1340,486 @@ int CLuaEngineDefs::EngineResetSurfaceProperties(lua_State* luaVM)
...
+{
+    //  bool engineSetModelPhysicalPropertiesGroup ( int modelID, int newGroup )
+    int          iModelID;
+    unsigned int iNewGroup;
+
+    CScriptArgReader argStream(luaVM);
+    argStream.ReadNumber(iModelID);
+    argStream.ReadNumber(iNewGroup);
+
+    if (!argStream.HasErrors())
```
Comment (patrikjuvonen) [inline Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp:1382]: These errors should be changed accordingly (https://github.com/multitheftauto/mtasa-blue/issues/821#issuecomment-522187565).

File: Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp [hunk 11]
```diff
@@ -1328,3 +1340,486 @@ int CLuaEngineDefs::EngineResetSurfaceProperties(lua_State* luaVM)
...
+
+int CLuaEngineDefs::EngineRestoreModelPhysicalPropertiesGroup(lua_State* luaVM)
+{
+    //  bool engineRestoreModelPhysicalPropertiesGroup ( int modelID )
+    int iModelID;
+
+    CScriptArgReader argStream(luaVM);
+    argStream.ReadNumber(iModelID);
+
+    if (!argStream.HasErrors())
```
Comment (patrikjuvonen) [inline Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp:1417]: These errors should be changed accordingly (https://github.com/multitheftauto/mtasa-blue/issues/821#issuecomment-522187565).

File: Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp [hunk 12]
```diff
@@ -1328,3 +1340,486 @@ int CLuaEngineDefs::EngineResetSurfaceProperties(lua_State* luaVM)
...
+{
+    //  bool engineSetObjectGroupPhysicalProperty ( int groupID, string property, ...)
+    int                      iGivenGroup;
+    eObjectGroup::Modifiable eProperty;
+
+    CScriptArgReader argStream(luaVM);
+    argStream.ReadNumber(iGivenGroup);
+    argStream.ReadEnumString(eProperty);
+
+    if (argStream.HasErrors())
```
Comment (patrikjuvonen) [inline Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp:1469]: These errors should be changed accordingly (https://github.com/multitheftauto/mtasa-blue/issues/821#issuecomment-522187565).

File: Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp [hunk 13]
```diff
@@ -1328,3 +1340,486 @@ int CLuaEngineDefs::EngineResetSurfaceProperties(lua_State* luaVM)
...
+{
+    //  bool engineGetObjectGroupPhysicalProperty ( int groupID, string property )
+    int                      iGivenGroup;
+    eObjectGroup::Modifiable eProperty;
+
+    CScriptArgReader argStream(luaVM);
+    argStream.ReadNumber(iGivenGroup);
+    argStream.ReadEnumString(eProperty);
+
+    if (argStream.HasErrors())
```
Comment (patrikjuvonen) [inline Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp:1652]: These errors should be changed accordingly (https://github.com/multitheftauto/mtasa-blue/issues/821#issuecomment-522187565).

File: Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp [hunk 14]
```diff
@@ -1328,3 +1340,486 @@ int CLuaEngineDefs::EngineResetSurfaceProperties(lua_State* luaVM)
...
+
+int CLuaEngineDefs::EngineRestoreObjectGroupPhysicalProperties(lua_State* luaVM)
+{
+    //  bool engineRestoreObjectGroupPhysicalProperties ( int groupID )
+    int iGivenGroup;
+
+    CScriptArgReader argStream(luaVM);
+    argStream.ReadNumber(iGivenGroup);
+
+    if (argStream.HasErrors())
```
Comment (patrikjuvonen) [inline Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp:1753]: These errors should be changed accordingly (https://github.com/multitheftauto/mtasa-blue/issues/821#issuecomment-522187565).

**Review by patrikjuvonen — CHANGES_REQUESTED** [review]
These errors should be changed accordingly (https://github.com/multitheftauto/mtasa-blue/issues/821#issuecomment-522187565).

**Comment by forkerer** [comment]
@patrikjuvonen Done :)

**Review by patrikjuvonen — APPROVED** [review]

**Comment by patrikjuvonen** [comment]
Great job! 👍 Thank you for the PR =)
